README.md
<div>
<img src="docs/images/logo.png" align="left" width="300"/>
<img src="docs/images/huygenslogo.png" align="right" width="200"/>
<img src="docs/images/tf.png" align="right" width="200"/>
<img src="docs/images/dans.png" align="right" width="100"/>
<img src="docs/images/huc.png" align="right" width="100"/>
</div>

# General Missives

[![SWH](https://archive.softwareheritage.org/badge/origin/https://github.com/Nino-cunei/oldassyrian/)](https://archive.softwareheritage.org/browse/origin/?origin_url=https://github.com/Dans-labs/clariah-gm)
[![DOI](https://zenodo.org/badge/292204502.svg)](https://zenodo.org/badge/latestdoi/292204502)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)


# Corpus

This repo contains a structurally clean version of the data of the *General Missives*, volumes 1-14.

Read more in [about](docs/about.md).

# Rationale for this representation of the corpus

Cleaning a textual dataset is a lot of work.
If such a dataset is a standard work, it will be studied by many students/researchers from several
disciplines. 
To make life easier for those people, they should be able to start with a dataset that is readily
processable by any tool of their choice.

[Text-Fabric](https://github.com/annotation/text-fabric)
provides a
[data model](https://annotation.github.io/text-fabric/tf/about/datamodel.html)
that captures the data at the end of the cleaning process just
before it goes into other tools.
It also support the integration of subsequent annotations with the original data.

The Missives corpus is an example how that works.

# Getting started

## Search interface to-go

For a first impression, start with
[missieven-search](https://CLARIAH.github.io/wp6-missieven-search/)
This is a static website that sends the whole corpus to your browser.
After a few seconds you can start searching.

You can do full text search via regular expressions, not only in the text,
but also in some of its attributes. For example, you can search for a word
in original letter texts or in editorial remarks.

More info in the
[manual](https://annotation.github.io/text-fabric/tf/about/clientmanual.html).

An example search is in [example.json](example.json).
Download the file, then import it in your search interface, and you see it happening.

You can save search results to excel files.

## Text-fabric browser

You get more power when you download Text-Fabric.
Text-Fabric operates in the ecosystem of Python and its libraries.

But you do not have to program in order to browse and search the corpus.
After installing Python and

```
pip3 install text-fabric
```

on the command line, say

```
text-fabric clariah/wp6-missieven
```

and a web server on your computer is started which serves you a search-and-browse
interface on the Generale Missives corpus.
You can search more precisely here than in the search interface-to-go above.

You can save search results to excel files.

## Jupyter notebooks

Text-Fabric is particularly suited to Jupyter notebooks.
There is a
[handy way to install](https://annotation.github.io/text-fabric/tf/about/install.html)
Python, JupyterLab in one go and Text-Fabric from there.

The next step is to consult the 
[tutorial](https://nbviewer.jupyter.org/github/CLARIAH/wp6-missieven/blob/master/tutorial/start.ipynb).
This is a series of notebooks that guides you to the computing facilities of Text-Fabric.
Text-Fabric is just a library that you import in your own Python programs,
which means that you can invoke the whole of Python and its libraries to do your job.
The only thing Text-Fabric does is to offer you a handy computing interface to the
textual data and their annotations.

See
[other corpora](https://annotation.github.io/text-fabric/tf/about/corpora.html)
for experiences with Text-Fabric as a pre-processing tool in other corpora.


# Getting the corpus data

The data of the corpus is in the `wp6-missieven` repo on GitHub:

* as simple, TEI-like XML (see the `xml` directory in this repo)
* as plain text-fabric files (see the `tf` directory in this repo)

If you use any method of working with the corpus indicated above, you do not have to
do anything special to download the data.
If you tell Text-Fabric it is in `clariah/wp6-missieven`,
it can find it and download it when needed. Automatically.


# Authors

This repo is by

*   [Dirk Roorda](https://github.com/dirkroorda) at
    [KNAW/HuC](https://huc.knaw.nl/di/text/)

## Acknowledgements

* Jesse de Does provided TEI-XML files for volumes 1-13.
* Lodewijk Petram provided textual PDFs for volume 14, bands (i) and (ii).
* Sophie Arnoult used the Text-Fabric data to perform Named Entity Recognition
  and delivered the results back.

# Status

*   2022-10-13 updates in Sophie's named entities, improved tutorials, source texts plus entities
    delivered as valid TEI-XML (see the [convert](https://nbviewer.org/github/CLARIAH/wp6-missieven/blob/master/programs/convert.ipynb) notebook)
*   2022-09-08 a tutorial notebook is added to show how to use the named entities
    detected by the algorithm of Sophie.
*   2022-05-04 version 1.0: Additional volumes: Volume 14, bands (i) and (ii) have been added.
    The earlier corrections by Sophie have not been re-applied, but the conversion has been improved 
    so that they are not needed anymore.
*   2022-04-11 Additional volumes: Volume 14, bands (i) and (ii) are in the process of being converted from
    textual PDF to Text-Fabric. Most structure has been recognized, but no TF has been generated yet.

[older ...](docs/history.md)


# Long term preservation and reproducibility

This repo has been archived in two independent places:

* at [Zenodo](https://zenodo.org)
* at [Software Heritage](https://archive.softwareheritage.org)

Click the respective badges above to be taken to the archives.
There you find ways to cite this work.

You can rerun the conversion programs on the source data and
regenerate the simple XML and Text-Fabric versions of the data.
See the
[reproduce](https://github.com/CLARIAH/wp6-missieven/blob/master/docs/reproduce.md).
guide.

# More interfaces
Another version of the data (less cleaned) is visible online in a
[BlackLab interface ](http://corpora.ato.ivdnt.org/corpus-frontend/Missiven/search)

A latent wish is to make the data of this repository available in a BlackLab interface.
In this repo we show how to set up a local BlackLab server and front-end and how to get the
present data into BlackLab.

**This is work in progress**, at this point follow the
[BlackLab install guide for macos](https://github.com/CLARIAH/wp6-missieven/blob/master/blacklab/install.md).

Thanks to Jesse de Does (key user of BlackLab, INT) and
Jan Niestadt (main author of BlackLab, INT) for helping out with setting up and using BlackLab.


docs/history.md
*   2021-07-22 Additional data corrections: letters that have no page break elements were
    not part of a page. That has been remedied (0.9), with the earlier corrections by Sophie on top of it (0.9.1).
*   2021-06-17 Data corrections by Sophie Arnoult have been applied (0.8.1)
*   2021-05-20 A new TF version (0.8) has been delivered.
    When multiple letters occur on one page, the words of the first line
    of some letters are not contained in line nodes.
    This has been corrected.
*   2021-01-30 A new TF version (0.7) has been delivered.
    This  version as a major encoding difference: whereas in version 0.6 footnote material ended
    up in the values of a feature, now footnotes are treated like text material.
    That means: the words in footnotes occupy slots, the footnotes themselves are nodes.
    As a consequence, footnotes and the words in it can be annotated, e.g. with named entities.
    This is a requirement for the kind of processing that Sophie Arnoult is currently devising.
*   2020-12-07 A new TF version (0.6) has been delivered.
    Fixed some folio references.
    Also: a simple data export has been made: a CSV file with all the words, and for each
    word whether it is editorial or original, and if a word has a footnote, the footnote is
    also given. See the
    [export notebook](https://nbviewer.jupyter.org/github/Dans-labs/clariah-gm/tree/master/usage/SimpleData.ipynb).
    The
    [exported data](https://github.com/Dans-labs/clariah-gm/releases/download/v0.6/words.tsv.gz)
    is attached to this release.
*   2020-11-24 Experimenting with a [Blacklab](http://inl.github.io/BlackLab/index.html) interface.
*   2020-11-17 A new TF version (0.5) has been delivered
    Fixed the generation of spurious newlines in footnote bodies.
*   2020-11-16 A new TF version (0.4) has been delivered
    Footnote bodies and marks have been checked and corrected, all encoded footnote marks
    have been linked to all encoded footnote bodies.
    Docs have been updated, and tutorials have been written.
*   2020-10-13 A new TF version (0.3) has been delivered
    Footnote bodies are almost all checked and corrected (12247 in total),
    footnote marks have been checked
    en corrected for volumes 1-4, there remain at least (300) pages with unlinked footnotes
    out of the 5270 pages that have footnotes.
    Editorial text is now in the main text, on equal footing with the original letter content,
    but separable from it in a number of ways.
*   2020-10-13 A new TF version (0.2) has been delivered, and there is now a TF-app
    [missieven](https://github.com/annotation/app-missieven) for this corpus.
    That means that functions like the Text-Fabric browser and easy downloading of data are supported.
    There is still cleaning work to do, especially in linking the footnotes to the proper
    footnote references.
    There are also a few mis-encoded tables (from landscape format), that need manual adjustment,
    and some pages that are altogether missing.
    See [trimTei0.py](https://github.com/Dans-labs/clariah-gm/blob/master/programs/trimTei0.py) where some of those
    pages have already been added.
*   2020-10-07 Many checks have been performed, many structural corrections
    w.r.t the TEI source have been performed,
    the metadata of all metadata has been thoroughly checked and corrected.
    See the reports in
    [trimreport2](trimreport2).
*   2020-09-16 First TF dataset created, but incomplete (notes are left out, checks needed)
    See 
    [last `trimTei` run](log-trimTei.txt)
    and
    [last `tfFromTrim` run](log-tfFromTrim.txt)
*   2020-09-02 Repository created, no content yet, start of conversion coding.





programs/lib.py
import sys
import os
import collections
import re
import yaml
import xml.etree.ElementTree as ET


# CONFIG READING


def readYaml(fileName):
    if os.path.exists(fileName):
        with open(fileName) as y:
            y = yaml.load(y, Loader=yaml.FullLoader)
    else:
        y = {}
    return y


# LOCATIONS

BASE = os.path.expanduser("~/github")
ORG = "clariah"
REPO = "wp6-missieven"

REPO_DIR = f"{BASE}/{ORG}/{REPO}"
LOCAL_DIR = f"{REPO_DIR}/_local"
SOURCE_DIR = f"{REPO_DIR}/source"
VERSION_PDF = "2022-03-09"
PDF_DIR = f"{SOURCE_DIR}/{VERSION_PDF}"
PDF_REPORT_DIR = f"{REPO_DIR}/pdfreport/{VERSION_PDF}"
TEST_DIR = f"{LOCAL_DIR}/test"
TEXT_DIR = f"{LOCAL_DIR}/text/{VERSION_PDF}"
TEXT_FILE = f"{PDF_REPORT_DIR}/text.txt"
FMT_FILE = f"{PDF_REPORT_DIR}/format.txt"
STRUCT_FILE = f"{PDF_REPORT_DIR}/struct.txt"
FINE_FILE = f"{PDF_REPORT_DIR}/fine.txt"
HEAD_FILE = f"{PDF_REPORT_DIR}/head.txt"
TIT_FILE = f"{PDF_REPORT_DIR}/titles.txt"
FN_BODY_FILE = f"{PDF_REPORT_DIR}/footnotebodies.txt"
FN_MARK_FILE = f"{PDF_REPORT_DIR}/footnotemarks.txt"

DECL_PATH = f"{REPO_DIR}/yaml"
META_DECL_FILE = f"{DECL_PATH}/meta.yaml"
META_DECL_FILE06 = f"{DECL_PATH}/meta06.yaml"

META_DECL = readYaml(META_DECL_FILE)
META_DECL06 = readYaml(META_DECL_FILE06)

VERSION_SRC = META_DECL["versionSrc"]
VERSION_TF = META_DECL["versionTf"]

IN_DIR = f"{SOURCE_DIR}/{VERSION_SRC}/tei"
TRIM_DIR = f"{LOCAL_DIR}/{VERSION_SRC}/trim"
CHANGE_DIR = f"{REPO_DIR}/change"
REPORT_DIR = f"{REPO_DIR}/trimreport"
XML_DIR = f"{REPO_DIR}/xml"
XMLOUT_DIR = f"{REPO_DIR}/xmlout"

E_ORG = "cltl"
E_REPO = "voc-missives"
# E_RELATIVE = "export/tf"
E_RELATIVE = "voc-missives/export/tf"
# ENTITY_TF = f"{BASE}/{E_ORG}/{E_REPO}/{E_RELATIVE}"
ENTITY_TF = f"{BASE}/{ORG}/{REPO}/{E_RELATIVE}"
# ENTITY_MOD = f"{E_ORG}/{E_REPO}/{E_RELATIVE}"
ENTITY_MOD = f"{ORG}/{REPO}/{E_RELATIVE}"

LAST_STAGE = 4

TF_DIR = f"{REPO_DIR}/tf"
OUT_DIR = f"{TF_DIR}/{VERSION_TF}"
PDF_VOL = "14"
BAND_OFFSET = {"1": 13, "2": -527}
RESUME_TABLE = 244

CELL_RE = re.compile(r"""<cell>(.*?)</cell>""", re.S)

LT = "≤"
GT = "≥"
AMP = "∪"

ADD_LB_ELEMENTS = set(
    """
    folio
    head
    para
    remark
    row
    subhead
""".strip().split()
)

ENSURE_LB_ELEMENTS = set(
    """
    head
""".strip().split()
)

A2Z = "abcdefghijklmnopqrstuvwxyz"


def ucFirst(word):
    if word is None or word == "":
        return word
    return f"{word[0].upper()}{word[1:]}"


def parseArgs(args):
    vol = None
    lid = None

    kwargs = {}
    pargs = []

    good = True

    for arg in args:
        if arg.isdigit() or "-" in arg:
            if "-" in arg:
                (b, e) = arg.split("-", 1)
                if b.isdigit() and e.isdigit():
                    values = set(range(int(b), int(e) + 1))
                else:
                    print(f"Unrecognized argument `{arg}`")
                    good = False
                    continue
            else:
                values = {int(arg)}
            if vol is None:
                vol = values
            elif lid is None:
                lid = values
        else:
            kv = arg.split("=", 1)
            if len(kv) == 1:
                pargs.append(arg)
            else:
                (k, v) = kv
                if k == "orig":
                    v = set(v.split(","))
                kwargs[k] = v

    if vol is not None:
        vol = {f"{i:>02}" for i in vol}
    if lid is not None:
        lid = {f"p{i:>04}" for i in lid}

    return (good, vol, lid, kwargs, pargs)


# FILESYSTEM OPERATIONS


def summarize(text, limit=20):
    lText = len(text)
    if lText <= limit:
        start = text
        inter = ""
        end = ""
    elif lText <= 2 * limit:
        start = text[0:limit]
        inter = ""
        end = text[limit:]
    else:
        start = text[0:limit]
        inter = " ... "
        end = text[-limit:]

    summary = f"{start:<{limit}}{inter:<5}{end:>{limit}}"
    trimmed = f"{start}{inter}{end}"

    return (summary, trimmed)


def docSummary(docs):
    nDocs = len(docs)
    rep = "   0x" if not nDocs else f"   1x {docs[0]}" if nDocs == 1 else ""
    if not rep:
        examples = " ".join(docs[0:2])
        rest = " ..." if nDocs > 2 else ""
        rep = f"{nDocs:>4}x {examples}{rest}"
    return f"{rep:<30}"


def clearTree(path):
    """Remove all files from a directory, recursively, but leave subdirectories.

    Reason: we want to inspect output in an editor.
    But if we remove the directories, the editor looses its current directory
    all the time.
    """

    subdirs = []
    with os.scandir(path) as dh:
        for (i, entry) in enumerate(dh):
            name = entry.name
            if name.startswith("."):
                continue
            if entry.is_file():
                os.remove(f"{path}/{name}")
            elif entry.is_dir():
                subdirs.append(name)

    for subdir in subdirs:
        clearTree(f"{path}/{subdir}")
        if subdir == "rest":
            os.rmdir(f"{path}/{subdir}")


def initTree(path, fresh=False):
    exists = os.path.exists(path)
    if fresh:
        if exists:
            clearTree(path)

    if not exists:
        os.makedirs(path, exist_ok=True)


# SOURCE READING


META_PAGE_NUM_RE = re.compile(
    r"""<interpGrp type="page"><interp>([0-9]+)</interp></interpGrp>""", re.S
)


def getPage(path):
    with open(path) as fh:
        text = fh.read()
    match = META_PAGE_NUM_RE.search(text)
    return f"p{int(match.group(1)):>04}" if match else None


def getVolumes(srcDir):
    volumes = []
    with os.scandir(srcDir) as dh:
        for entry in dh:
            if entry.is_dir():
                vol = entry.name
                if not vol.isdigit():
                    continue
                volumes.append(f"{int(vol):>02}")
    return sorted(volumes)


def getLetters(srcDir, idMap=None):
    letters = []
    collisions = 0
    lidSet = set()

    with os.scandir(srcDir) as dh:
        for entry in dh:
            if entry.is_file():
                name = entry.name
                if name.endswith(".xml"):
                    pureName = name[0:-4]
                    if idMap is not None:
                        lid = getPage(f"{srcDir}/{name}")
                        if lid in lidSet:
                            collisions += 1
                        lidSet.add(lid)
                        idMap[pureName] = lid
                    letters.append(pureName)
    if collisions:
        print(f"MAPPING from letter ids to first pages: {collisions} collisions")
        return ()
    return tuple(sorted(letters))


# ANALYSIS


NUM_RE = re.compile(r"""[0-9]""", re.S)


def nodeInfo(node, analysis, after=False):
    tag = node.tag
    atts = node.attrib

    if not atts:
        analysis[f"{tag}.."] += 1
    else:
        for (k, v) in atts.items():
            vTrim = NUM_RE.sub("n", v)
            if k == "value":
                vTrim = "x"
            isLayout = k == "rend"
            if after:
                if isLayout:
                    analysis[f" {tag}.{k}={v}"] += 1
                else:
                    analysis[f"{tag}.{k}={vTrim}"] += 1
            else:
                if not isLayout:
                    analysis[f"{tag}.{k}={vTrim}"] += 1

    for child in node:
        nodeInfo(child, analysis, after=after)


def analyse(text, after=False):
    analysis = collections.Counter()
    root = ET.fromstring(text)
    nodeInfo(root, analysis, after=after)
    return analysis


# TOP-LEVEL TRIM


def trim(
    stage,
    givenVol,
    givenLid,
    corpusPre,
    trimVolume,
    trimDocBefore,
    trimDocPrep,
    trimPage,
    processPage,
    trimDocPost,
    corpusPost,
    *args,
    **kwargs,
):
    if corpusPre:
        corpusPre(givenVol)

    SRC = IN_DIR if stage == 0 else f"{TRIM_DIR}{stage - 1}"
    DST = XML_DIR if stage == LAST_STAGE else f"{TRIM_DIR}{stage}"
    REP = f"{REPORT_DIR}{stage}"
    initTree(REP)

    volumes = getVolumes(SRC)

    analysis = collections.Counter()
    analysisAfter = collections.Counter()

    info = dict(
        table=0,
        tableDiag={},
        docs=[],
        pageDiag=collections.defaultdict(dict),
        metas=0,
        metasUnknown=[],
        metasDistilled=collections.defaultdict(dict),
        nameDiag=collections.defaultdict(
            lambda: collections.defaultdict(lambda: collections.defaultdict(list))
        ),
        metaValues=collections.defaultdict(
            lambda: collections.defaultdict(lambda: collections.defaultdict(list))
        ),
        metaDiag=collections.defaultdict(dict),
        captionInfo=collections.defaultdict(list),
        captionNorm=collections.defaultdict(list),
        captionVariant=collections.defaultdict(list),
        captionRoman=collections.defaultdict(list),
        firstP={},
        folioResult=collections.defaultdict(list),
        folioTrue=collections.defaultdict(list),
        folioFalse=collections.defaultdict(list),
        folioUndecided=collections.defaultdict(lambda: collections.defaultdict(list)),
        headInfo=collections.defaultdict(list),
        remarks=collections.Counter(),
        remarkInfo=collections.defaultdict(lambda: collections.defaultdict(list)),
        fnoteBodyInfo=collections.defaultdict(list),
        fnoteMarkInfo=collections.defaultdict(list),
        heads={},
        bigTitle={},
        splits=[],
        splitsX=[],
    )

    mergeText = {}
    previousDoc = None

    for vol in volumes:
        if givenVol is not None and vol not in givenVol:
            continue
        print(f"\rvolume {vol}" + " " * 70)

        volDir = vol.lstrip("0") if stage == 0 else vol
        thisSrcDir = f"{SRC}/{volDir}"
        thisDstDir = f"{DST}/{vol}"
        rest = "/rest" if stage == 1 else ""
        initTree(thisDstDir, fresh=True)
        initTree(f"{thisDstDir}{rest}", fresh=True)

        info["vol"] = vol
        idMap = {} if stage == 0 else None
        letters = getLetters(thisSrcDir, idMap=idMap)

        if trimVolume:
            trimVolume(vol, letters, info, idMap, givenLid, mergeText)

        for name in letters:
            lid = name if idMap is None else idMap[name]
            if givenLid is not None and lid.rstrip(A2Z) not in givenLid:
                continue
            doc = f"{vol}:{lid}"

            if trimDocBefore is None:
                with open(f"{thisSrcDir}/{name}.xml") as fh:
                    text = fh.read()
            else:
                text = trimDocBefore(doc, name, thisSrcDir, mergeText)
                if text is None:
                    continue

            sys.stderr.write(f"\r\t{lid}      ")
            info["doc"] = doc
            info["docs"].append(doc)

            origText = text

            thisAnalysis = analyse(text)
            for (path, count) in thisAnalysis.items():
                analysis[path] += count
            text = trimDocument(
                stage,
                text,
                trimPage,
                info,
                processPage,
                trimDocPrep,
                trimDocPost,
                previousDoc,
                *args,
                **kwargs,
            )
            if text is None:
                dstName = f"{lid}.xml"
                with open(f"{thisDstDir}/rest/{dstName}", "w") as fh:
                    fh.write(origText)
                continue

            dstName = f"{lid}.xml"
            with open(f"{thisDstDir}/{dstName}", "w") as fh:
                fh.write(text)

            thisAnalysisAfter = analyse(text, after=True)
            for (path, count) in thisAnalysisAfter.items():
                analysisAfter[path] += count

            previousDoc = doc

    print("\rdone" + " " * 70)

    docs = info["docs"]
    totalDocs = len(docs)
    print(f"{totalDocs} documents")

    with open(f"{REP}/elementsIn.tsv", "w") as fh:
        for (path, amount) in sorted(analysis.items()):
            fh.write(f"{path}\t{amount}\n")

    with open(f"{REP}/elementsOut.tsv", "w") as fh:
        for (path, amount) in sorted(analysisAfter.items()):
            fh.write(f"{path}\t{amount}\n")

    if corpusPost:
        corpusPost(info)


BODY_RE = re.compile(r"""<body[^>]*>(.*?)</body>""", re.S)


BODY_RE = re.compile(r"""<body[^>]*>(.*?)</body>""", re.S)


def trimDocument(
    stage,
    text,
    trimPage,
    info,
    processPage,
    trimDocPrep,
    trimDocPost,
    previousDoc,
    *args,
    **kwargs,
):
    headElem = "teiHeader" if stage == 0 else "header"
    headerRe = re.compile(rf"""<{headElem}[^>]*>(.*?)</{headElem}>""", re.S)
    match = headerRe.search(text)
    metaText = match.group(1)
    match = BODY_RE.search(text)
    bodyText = match.group(1)

    if trimDocPrep:
        (header, bodyText) = trimDocPrep(info, metaText, bodyText, previousMeta)
    else:
        header = f"""<header>\n{metaText}\n</header>"""

    if bodyText is None:
        return None

    body = trimBody(stage, bodyText, trimPage, info, processPage, *args, **kwargs)

    if trimDocPost:
        body = trimDocPost(info, body)

    return (
        f"""<teiTrim>\n{header}\n<body>\n{body}\n</body>\n</teiTrim>"""
        if body
        else None
    )


# TRIM HEADER


WHITE_NLNL_RE = re.compile(r"""\n{3,}""", re.S)
WHITE_NL_RE = re.compile(r"""(?:[ \t]*\n[ \t\n]*)""", re.S)
SPACE_RE = re.compile(r"""  +""", re.S)
WHITE_RE = re.compile(r"""\s\s+""", re.S)
PB_RE = re.compile(r"""<pb\b[^>]*/>""", re.S)


previousMeta = {}


PB_CHECK_PAT = "|".join(
    """
    fw
    head
    hi
    table
    cell
    row
    note
""".strip().split()
)
CHECK_PB_RE = re.compile(rf"""<({PB_CHECK_PAT})\b[^>]*>(.*?)</\1>""", re.S)


def checkPb(text):
    breaks = collections.Counter()

    for match in CHECK_PB_RE.finditer(text):
        elem = match.group(1)
        material = match.group(2)
        if "<pb" in material:
            breaks[elem] += 1
    return breaks


X_RE = re.compile(r"""\s*xml:?[a-z]+=['"][^'"]*['"]""", re.S)
PB_FACS_TPL_RE = re.compile(r'''\b(?:facs|tpl)="[^"]*"''', re.S)
PAGE_NUM_RE = re.compile(r"""<pb[^>]*?\bn=['"]([0-9]+)['"]""", re.S)

NL_B_RE = re.compile(r"""\n*<(p|para|note|fw|table|row)\b([^>]*)>""", re.S)
NL_E_RE = re.compile(r"""</(p|para|note|fw|table|row)\b>\n*""", re.S)
LB_RE = re.compile(r"""<lb/>\n*""", re.S)
DIV_CLEAN_RE = re.compile(r"""</?div\b[^>]*>""", re.S)


def trimBody(stage, text, trimPage, info, processPage, *args, **kwargs):
    if stage == 0:
        breaks = checkPb(text)
        if breaks:
            print("\nsensitive page breaks")
            for (elem, n) in sorted(breaks.items()):
                print(f"\t{n:>3} x {elem}")

        text = DIV_CLEAN_RE.sub(r"", text)

    prevMatch = 0
    previous = {}
    result = []
    first = True

    def doPage(page, *args, **kwargs):
        nonlocal first

        if page is None:
            if processPage is not None:
                processPage(None, previous, result, info, *args, **kwargs)
                first = False
            return

        match = PAGE_NUM_RE.search(page)
        pageNum = f"-{match.group(1):>04}" if match else ""
        info["page"] = f"{info['doc']}{pageNum}"
        info["pageNum"] = pageNum.lstrip("-")
        info["first"] = first
        first = False
        if stage == 0:
            page = X_RE.sub("", page)
            page = PB_FACS_TPL_RE.sub("", page)
        if processPage is None:
            page = trimPage(page, info, *args, **kwargs)
            page = LB_RE.sub(r"""<lb/>\n""", page)
            page = PB_RE.sub(r"""\n\n\g<0>\n\n""", page)
            page = NL_B_RE.sub(r"""\n<\1\2>""", page)
            page = NL_E_RE.sub(r"""\n</\1>\n""", page)
            if stage == 0:
                page = page.replace(" <hi", "\n<hi")
            page = WHITE_NL_RE.sub("\n", page.strip())
            page = SPACE_RE.sub(" ", page)
            result.append(page)
        else:
            processPage(page, previous, result, info, *args, **kwargs)

    for match in PB_RE.finditer(text):
        b = match.start()
        thisPage = text[prevMatch:b].strip()
        if not thisPage:
            continue
        doPage(thisPage, *args, **kwargs)
        result.append("\n")
        prevMatch = b

    if prevMatch < len(text):
        thisPage = text[prevMatch:].strip()
        doPage(thisPage, *args, **kwargs)
    doPage(None, *args, kwargs)

    body = "".join(result)

    match = PB_RE.search(body)
    prePage = body[0 : match.start()].strip() if match else ""
    if prePage:
        print(f"\nMaterial before first page\n\t=={prePage}==")

    body = WHITE_NLNL_RE.sub("\n\n", body.strip())
    return body


def applyCorrections(corrections, doc, text):
    if doc in corrections:
        for (correctRe, correctRepl, correctN) in corrections[doc]:
            (text, n) = correctRe.subn(correctRepl, text)
            if n < correctN:
                print(text)
                print(
                    f"\tCORRECTION {doc} {correctRe.pattern} {correctN}"
                    f" under-applied ({n}x)"
                )
            elif n > correctN:
                print(text)
                print(
                    f"\tCORRECTION {doc} {correctRe.pattern} {correctN}"
                    f" over-applied ({n}x)"
                )
    return text


def rangesFromList(nodeList):  # the list must be sorted
    curstart = None
    curend = None
    for n in nodeList:
        if curstart is None:
            curstart = n
            curend = n
        elif n == curend + 1:
            curend = n
        else:
            yield (curstart, curend)
            curstart = n
            curend = n
    if curstart is not None:
        yield (curstart, curend)


def specFromRanges(ranges):  # ranges must be normalized
    return ",".join(
        "{}".format(r[0]) if r[0] == r[1] else "{}-{}".format(*r) for r in ranges
    )
